File: go/stocks/portfolio.go
package stocks

// さまざまな currency の Money を要素として持つスライス
type Portfolio []Money

func (p Portfolio) Add(money Money) Portfolio {
	p = append(p, money)
	return p
}

func (p Portfolio) Evaluate(currency string) Money {
	// return Money{amount: 15, currency: "USD"}
	total := 0.0
	for _, m := range p {
		total = total + convert(m, currency)
	}
	return Money{amount: total, currency: currency} // 返り値の currency はEvaluate メソッドの第一引数にのみ依存しているため、この実装では不十分である
}

// money と 変換する通貨を選択すると、為替レートを使って変換を行う
func convert(money Money, currency string) float64 {
	// 為替レートを保存した hash map
	exchangeRates := map[string]float64{
		"EUR->USD": 1.2,
		"USD->KRW": 1100,
	}

	if money.currency == currency {
		return money.amount
	}

	key := money.currency + "->" + currency // money.currency と 　convertの引数にとった currency を用いて key を作る

	return money.amount * exchangeRates[key] // key を使って、exchangeRates に登録してある為替レートを取り出す
}

File: js/portfolio.js
const Money = require('./money'); // Prtofolio モジュールは Money モジュールに依存性があるため require しなければ ならない
class Portfolio{

    constructor(){
        this.moneys = [];
    }

    add(...moneys){ // money を複数引数としてとることを表す. rest parameter syntax
        this.moneys = this.moneys.concat(moneys); // 引数にとった money の列を連結して,moneys フィールドに格納する
    }

    evaluate(currency) {
        // return new Money(15, "USD"); // テストを通すためだけにハードコードしている(リファクタの段階で修正)
        let total = this.moneys.reduce((sum, money) => { // reduce は配列 moneys に対して定義されている組み込みメソッドで、sum の初期値を 0 として、moneysの各要素の money を用いて計算を"畳み込んでいく". 関数型でいうと foldleft  
            return sum + this.convert(money, currency); // 配列 moneys の 要素のmoney は evaluate の引数として与えられた通貨に対応して,convertされてから、足されていく
        }, 0);
        return new Money(total, currency); // ここで、Money モジュールに依存している
    }

    convert(money, currency) {
        // 為替レートを保存するマップを用意
        let exchangeRates = new Map();
        exchangeRates.set("EUR->USD", 1.2);
        exchangeRates.set("USD->KRW", 1100);
        if(money.currency === currency) {
            return money.amount;
        }
        let key = money.currency + "->" + currency;
        return money.amount * exchangeRates.get(key);
    }
}

module.exports = Portfolio;